File: CppSource/Source/System/user_impl.cpp
// Copyright (c) Microsoft Corporation
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

#include "pch.h"
#include "user_impl.h"

XboxLiveUserImpl::XboxLiveUserImpl(
    _In_ Windows::System::User^ creationContext,
    _In_ XboxLiveUser *cUser
    )
    : m_cUser(cUser)
{
    if (creationContext != nullptr)
    {
        m_cppUser = std::make_shared<xbox::services::system::xbox_live_user>(creationContext);
    }
    else
    {
        m_cppUser = std::make_shared<xbox::services::system::xbox_live_user>();
    }
}

void XboxLiveUserImpl::Refresh()
{
    if (m_cUser != nullptr)
    {
        m_xboxUserId = m_cppUser->xbox_user_id();
        m_cUser->xboxUserId = m_xboxUserId.data();

        m_gamertag = m_cppUser->gamertag();
        m_cUser->gamertag = m_gamertag.data();

        m_ageGroup = m_cppUser->age_group();
        m_cUser->ageGroup = m_ageGroup.data();

        m_privileges = m_cppUser->privileges();
        m_cUser->privileges = m_privileges.data();

        m_cUser->isSignedIn = m_cppUser->is_signed_in();

#if UWP_API
        m_webAccountId = m_cppUser->web_account_id();
        m_cUser->webAccountId = m_webAccountId.data();
#endif
        m_cUser->windowsSystemUser = m_cppUser->windows_system_user();
    }
}

File: CSharpSource/Source/Microsoft.Xbox.Services.UnityXDK.CSharp/System/UserImpl.cs
// Copyright (c) Microsoft Corporation
// Licensed under the MIT license. See LICENSE file in the project root for full license information.

namespace Microsoft.Xbox.Services.System
{
    using global::System;
    using global::System.Runtime.InteropServices;
    using global::System.Threading.Tasks;

    internal class UserImpl : IUserImpl
    {
        public bool IsSignedIn { get; set; }
        public XboxLiveUser User { get; set; }

        public string XboxUserId { get; set; }
        public string Gamertag { get; set; }
        public string AgeGroup { get; set; }
        public string Privileges { get; set; }
        public string WebAccountId { get; set; }
        public AuthConfig AuthConfig { get; set; }

        public IntPtr m_xboxLiveUser_c;
        private Windows.Xbox.System.User m_xboxSystemUser;

        public UserImpl(Windows.Xbox.System.User xboxSystemUser)
        {
            m_xboxLiveUser_c = XboxLive.Instance.Invoke<IntPtr, XboxLiveUserCreate>(null);
        }

        public Task<SignInResult> SignInImpl(bool showUI, bool forceRefresh)
        {
            throw new NotImplementedException();
        }

        public Task<TokenAndSignatureResult> InternalGetTokenAndSignatureAsync(string httpMethod, string url, string headers, byte[] body, bool promptForCredentialsIfNeeded, bool forceRefresh)
        {
            return Task.Run(() =>
            {
                var getTokenAndSignatureResult = m_xboxSystemUser.GetTokenAndSignatureAsync(httpMethod, url, headers, body).GetResults();

                return new TokenAndSignatureResult
                {
                    // TODO populate other fields
                    Token = getTokenAndSignatureResult.Token,
                    Signature = getTokenAndSignatureResult.Signature
                };
            });
        }

        private delegate IntPtr XboxLiveUserCreate(IntPtr xboxSystemUser);

        [StructLayout(LayoutKind.Sequential)]
        private struct XboxLiveUser_c
        {
            [MarshalAsAttribute(UnmanagedType.LPWStr)]
            public string XboxUserId;

            [MarshalAsAttribute(UnmanagedType.LPWStr)]
            public string Gamertag;

            [MarshalAsAttribute(UnmanagedType.LPWStr)]
            public string AgeGroup;

            [MarshalAsAttribute(UnmanagedType.LPWStr)]
            public string Privileges;

            [MarshalAsAttribute(UnmanagedType.U1)]
            public byte IsSignedIn;

            [MarshalAsAttribute(UnmanagedType.SysInt)]
            public IntPtr XboxSystemUser;
        };
    }
}